### ISSUE TITLE
Feature Request: support deep sleep for the RP2350

### ISSUE THREAD

**Issue by sillycowvalley** [earlephilhower/arduino-pico#2528]
I wouldn't know where to start so I asked the One Who Is Wrong With Confidence (ChatGPT) for an example and got the example below as a starting point. Would we want to add some RISC-V specific functionality to the RP2040 class for this?

---

Here's an example of how to put the Pico 2 into deep sleep and wake it up using a GPIO pin (You can also use the RTC to wake it up after a specific time interval):

### C++ Example for Deep Sleep on RP2350 (Pico 2)

```cpp
#include "pico/stdlib.h"
#include "hardware/gpio.h"
#include "hardware/clocks.h"
#include "hardware/vreg.h"
#include "hardware/regs/io_bank0.h"

// Set the GPIO pin to wake the system
#define WAKE_PIN 15

void setup_wake_pin() {
    // Enable the wake pin
    gpio_set_dir(WAKE_PIN, GPIO_IN);
    gpio_pull_down(WAKE_PIN);

    // Enable wakeup by pin
    gpio_set_irq_enabled(WAKE_PIN, GPIO_IRQ_EDGE_RISE, true);
}

int main() {
    // Optional: Lower the voltage to reduce power consumption in sleep
    vreg_set_voltage(VREG_VOLTAGE_0_90);

    // Initialize stdio, not needed for actual deep sleep but for testing/debugging
    stdio_init_all();

    // Configure the GPIO pin for waking up
    setup_wake_pin();

    // Print message before going to deep sleep
    printf("Going to sleep now...\n");

    // Enable deep sleep (Note: the Pico 2 allows real deep sleep with RTC wakeup)
    sleep_goto_dormant();

    // This point will only be reached when the system is woken up
    printf("Woke up from sleep!\n");

    return 0;
}
```

### Key Steps in the Example:
1. **GPIO Wake Pin**: Set a GPIO pin (e.g., pin 15) to wake up the device from deep sleep when a rising edge is detected.
2. **Voltage Reduction**: Optionally, you can reduce the core voltage to 0.9V using `vreg_set_voltage()` to further lower power consumption.
3. **GPIO Interrupts**: The code configures the GPIO interrupts to handle the wake-up.
4. **sleep_goto_dormant()**: This is the key function that puts the Pico into deep sleep, cutting power to most peripherals while maintaining the ability to wake up via external signals.

This example ensures the Pico 2 goes into a low-power deep sleep and wakes up based on a GPIO event.

**Comment by earlephilhower** [comment]
I'm pretty sure the code sample you got isn't close to correct, but there are good examples from the RPI folks directly in the `pico-examples` and the SDK and datasheet docs.

I still haven't found out where my USB power monitor migrated to so don't really have the equipment to do this myself.

If someone wants to give ti a go, I'd be happy to get it in here.  The interesting bit will be getting the system and core back into a consistent state after restart.  Things like which clocks are on, what the USB state is, which blocks are enabled (ADC, PWM, timer, etc.).  Requiring the user to assume the board is equivalent to power on (but with valid RAM) might be fine, but the core internal state will need to be updated to reflect whatever is chosen...

**Comment by maxgerhardt** [comment]
For the RP2350 there is at least the powman (power manager) API that can switch off certain things. And the CMSIS headers for the ARM Cortex-M33 and the RISC-V Hazard3 probably also have functions to go into a sleep state (wait for interrupt, etc.).

Also related to https://github.com/adafruit/circuitpython/issues/9491 and https://github.com/raspberrypi/pico-playground/pull/48

**Comment by seamusdemora** [comment]
I need x-low power consumption from a uC for an off-grid project I've undertaken. I'd like to use the Pico for this as I've done another project (controlling an INA260) using a Pico 1. _**However:**_ I have tried the [`dormant` example from the Pico Playground](https://github.com/raspberrypi/pico-playground/tree/master/sleep) on an Pico 1, and it simply did not work for me. I'm not sure it works at all... 

I'm posting to ask about the status of "deep sleep" here: does it appear feasible/do-able, any thoughts on the timeline??

My alternative for the Pico is a TI MSP430 - they have examples that actually work :)

**Comment by maxgerhardt** [comment]
I just tested some code found in the RP2040 datasheet and ran it on my Pico (1). The DORMANT sleep with  GPIO wakeup worked just fine.

Here's the sketch:
```cpp
#include <Arduino.h>
#include <hardware/xosc.h>

#define GPIO_EXIT_DORMANT_MODE 3 /* connect GP3 to GND once in DORMANT mode */

void setup() {
  Serial1.begin(9600);
  pinMode(LED_BUILTIN, OUTPUT);
  digitalWrite(LED_BUILTIN, HIGH);
  pinMode(3, INPUT_PULLUP); // pull pin that will get us out of sleep mode
}

void loop() {
  if (Serial1.available() > 0) {
    (void) Serial1.read();
    digitalWrite(LED_BUILTIN, LOW);
    // this interrupt source will get us out of the DORMANT mode agin
    gpio_set_dormant_irq_enabled(GPIO_EXIT_DORMANT_MODE, IO_BANK0_DORMANT_WAKE_INTE0_GPIO0_EDGE_LOW_BITS, true);
    // disable systick now so that no milisecond interrupts will occur
    systick_hw->csr &= ~1;
    // we will get out of sleep when an interrupt occurs.
    // this will shutdown the crystal oscillator until an interrupt occurs.
    xosc_dormant();
    // this will only be reached after wakeup.
    gpio_acknowledge_irq(GPIO_EXIT_DORMANT_MODE, IO_BANK0_DORMANT_WAKE_INTE0_GPIO0_EDGE_LOW_BITS); // acknowledge GPIO IRQ
    systick_hw->csr |= 1; // enable systick again, hope we survived this
    // we don't actually know the time duration during which we were dormant.
    // so, the absolute value ofmillis() will be messed up.
    digitalWrite(LED_BUILTIN, HIGH);
  }
  Serial1.println("hello, world");
  delay(500);
}
```
note that this was compiled with the USB stack **off**, so `Seriral1` is the hardware UART on GP0, GP1.

The sketch waits on a character to be received via UART, then it will set the crystal oscillator into dormant mode. Before that, GP3 is registered as a wakeup source, expecting a falling edge.

With a Nordic PPK2, I was able to measure the current going into the 5V "VBUS" line. So, this measures the **entire** current draw of the Pico board at 5V, and not just that of the RP2040 processor.

I can clearly see that, once I send the UART character, the current draw drops dramatially, from an average of 23.54mA to 2.10mA (at 5V).

![grafik](https://github.com/user-attachments/assets/84088687-92b9-4b3f-a5a4-b3f8bf4534b5)

**Comment by maxgerhardt** [comment]
The `pico_sleep` component was also recently updated: https://github.com/raspberrypi/pico-extras/tree/master/src/rp2_common/pico_sleep

**Comment by maxgerhardt** [comment]
Mhm, the Pico datasheet claims 0.8mA @ 5V in DORMANT mode ([source](https://datasheets.raspberrypi.com/pico/pico-datasheet.pdf)).. If I manually copy in the pico_sleep and hardware_rosc component from above (with a few mods to get rid of stdio reinit and enable a GPIO pullup) and do a 
```cpp
    sleep_run_from_dormant_source(DORMANT_SOURCE_XOSC);
    sleep_goto_dormant_until_pin(GPIO_EXIT_DORMANT_MODE, true, false); 
```
I do get down to 1.88mA. Something is still missing here.. maybe a perpiheral is on, or some other component on the board is drawing power. In any case, enough experiments for today.

![grafik](https://github.com/user-attachments/assets/466068e4-602c-42b6-a3f0-63b3beb1da49)

**Comment by earlephilhower** [comment]
Off the cuff, maybe try setting all GPIOs to `INPUT_PULLUP` to avoid wasting any pad power on floating pins?

Also probably want to `Serial1.end()` to put the UART0 into reset.  It's still a digital device running off of some clock even when idle, with input protection FFs/etc. always doing something...

**Comment by seamusdemora** [comment]
> I just tested some code found in the RP2040 datasheet and ran it on my Pico (1). The DORMANT sleep with GPIO wakeup worked just fine.
> 
> Here's the sketch:

@maxgerhardt : 

That's a big help - thank you! Uh... where is the "RP2040 datasheet" where you found this?

**Comment by maxgerhardt** [comment]
The pico_sleep component works on both RP2040 and RP2350 chips, it differentiates between them using `#if (!)PICO_RP2040` at compile time.

**Comment by seamusdemora** [comment]
@maxgerhardt 
> I just tested some code found in the RP2040 datasheet and ran it on my Pico (1). The DORMANT sleep with GPIO wakeup worked just fine. 

Excuse me please, but I still don't understand what you mean by "code found in the RP2040 datasheet".  Can you provide some clarification - or a link to that code? I have looked in the [Raspberry Pi Pico Datasheet](https://datasheets.raspberrypi.com/pico/pico-datasheet.pdf), but I find no code in there. 

I'm also confused by the source of the [code you posted in your earlier comment](https://github.com/earlephilhower/arduino-pico/issues/2528#issuecomment-2466951767)... It's not the same as this ['hello_dormant' example in the Pico Playground repo](https://github.com/raspberrypi/pico-playground/blob/master/sleep/hello_dormant/hello_dormant_gpio.c). Where do I find this code? 

My apologies for acting as a nincompoop, but I'm not really familiar at all with this `arduino-pico` repository. Can you provide me with a brief pointer to - or explanation - on how to get started with this??

**Comment by seamusdemora** [comment]
@maxgerhardt 

I've opened a discussion on an [issue I'm having getting your example above to compile](https://github.com/earlephilhower/arduino-pico/discussions/2601#discussioncomment-11234970)... I'd surely appreciate you taking a look at it.

**Comment by makejimmi** [comment]
Hi, 
try deinitializing the cyw43-module (cyw43_arch_deinit()) before calling sleep_run_from_dormant_source().
Reached ~380 uA @ 5V with both cores running on the Pico 2W using the same PPK2 measurement tool. Make sure to initialize it (cyw43_arch_init()) again after sleep_power_up() in order to restore all the Wireless functionalities.

As far as I know, the module gets initialized during the boot sequence before any core gets up and running which is why the module will always draw current. I would not be able to pinpoint the location in which it gets initialized but probably here: [variants init](https://github.com/earlephilhower/arduino-pico/blob/7c51742522b54b1dbf5492fe2b513461c3cf1246/variants/rpipico2w/init.cpp#L23).

This consequently comes with the disadvantage of having all Wireless functionalities completely stopped when going and staying DORMANT but if it is about saving as much power as possible then it surely is worth considering.

![Image](https://github.com/user-attachments/assets/c5e5275f-d664-4a9b-bc5a-99a150d8f6e0)

Settings: cyw43_arch_deinit, Brown-Out Detection Off, Pin 4 Pull Up as wake up source

**Comment by seamusdemora** [comment]
> Hi, try deinitializing the cyw43-module (cyw43_arch_deinit()) before calling sleep_run_from_dormant_source(). Reached ~380 uA @ 5V with both cores running on the Pico 2W using the same PPK2 measurement tool and the Low Power Oscillator as DORMANT source. Make sure to initialize it (cyw43_arch_init()) again after sleep_power_up() in order to restore all the Wireless functionalities.

@makejimmi : 

Thanks for the update... I'm fighting Linux audio battles at present, but will get back & try this at first opportunity; I'm sure I'll have more questions then.

**Comment by JessicaJones-BT** [comment]
I would love to see this for some battery-powered IOT projects I have kicking around in my head. Personally I have no use for WiFi or much of anything else, the projects I'm looking for are basically just offline/disconnected "sentry" devices that wake up periodically, check some sort of sensor, and then start alerting (light and/or sound) if there's a problem.

**Comment by adeldigital** [comment]
@makejimmi Did you do any testing to check for the minimum power consumption in dormant mode with SRAM retention?

**Comment by radek-turecki** [comment]
Is it possible to make RP2350 into deepsleep mode but for a time (not for pin wake-up)?

**Comment by dquadros** [comment]
> Is it possible to make RP2350 into deepsleep mode but for a time (not for pin wake-up)?

Yes,  Power Management (POWMAN) includes an alarm timer. See section 6.2 of the RP2350 datasheet and the hardware_powmen API documentation (https://www.raspberrypi.com/documentation/pico-sdk/hardware.html#group_hardware_powman). You need a clock source for the timer.

**Comment by radek-turecki** [comment]
@dquadros but hardware_powman as far as I try based on doc (https://www.raspberrypi.com/documentation/pico-sdk/hardware.html) it doesn't works when I try to call function from powman in Arduino eg.
`powman_timer_set_1khz_tick_source_lposc();`

I have #include "pico/stdlib.h" bat the beginnig of code of course, but during compilation I get the error - "**_Compilation error: 'powman_timer_set_1khz_tick_source_lposc' was not declared in this scope_**"

**Comment by maxgerhardt** [comment]
Are you doing
```cpp
#include "hardware/powman.h"
```
? See https://github.com/raspberrypi/pico-sdk/blob/2.1.1/src/rp2_common/hardware_powman/include/hardware/powman.h#L29-L32

**Comment by radek-turecki** [comment]
in the code:
#include <hardware/xosc.h>
#include "hardware/powman.h"

for hardware/xosc.h is OK

but for hardware/powman is error:
_**Compilation error: hardware/powman.h: No such file or directory**_

**Comment by maxgerhardt** [comment]
True, because the include path

```
-iwithprefixbefore/pico-sdk/src/rp2_common/hardware_powman/include
```

is missing from https://github.com/earlephilhower/arduino-pico/blob/master/lib/core_inc.txt currently.

**Comment by maxgerhardt** [comment]
I've uploaded my lowpower tests from previously to 

https://github.com/maxgerhardt/pico-lowpower-testing

you might want to use that and check out the 

```c
void sleep_goto_dormant_until(struct timespec *ts, aon_timer_alarm_handler_t callback);
```

API, that will allow you to sleep until a certain timestamp has been reached. The current example compiles for both RP2050 and RP2350 by duplicating the "missing" (or unreachable) header files into the project.

**Comment by radek-turecki** [comment]
Oooo, great! 

Do you know if the include path will be fixed?

I tried your code version with pin wake-up, i.e. these lines:
`sleep_run_from_dormant_source(DORMANT_SOURCE_XOSC);
 sleep_goto_dormant_until_pin(GPIO_EXIT_DORMANT_MODE, true, false); `
but the measured consumption was 10mA, which is too much for me, I need to go lower.

Thanks for everything, then. I'll dive into the literature, read on, and try out the function suggested by you (`void sleep_goto_dormant_until(struct timespec *ts, aon_timer_alarm_handler_t callback)`;)

**Comment by maxgerhardt** [comment]
Note that my example code waits for a character to be received via UART1 (at TX=GP0, RX=GP1, **not** the USB serial) before going to dormant mode, if you want to remove that just call theese functions directly. I've also only tested that code on RP2040 not yet on RP2350, but I do have my PPK2 and a board here, so let's see.. If you're not seeing a difference from a regular firmware execution then the sleep code is not executing correctly.

Of course, other things play a factor too, like other components on the board or the voltage applied to pins causing a leakage current (e.g., driving a pin low while an external pullup resistor is connected --> leakage current through the resistor). In general, for testing, I upload the code via SWD, then disconnect the SWD cables completely as to not cause a posisble leakage. The PPK2 powers either VBUS with 5V or VSYS with 3.3V (see "bypassing the regulator" comments above).

**Comment by maxgerhardt** [comment]
Missing access to powman.h will be fixed by https://github.com/earlephilhower/arduino-pico/pull/3399.

**Comment by radek-turecki** [comment]
hmmm but json in relesae is not updated?
`https://github.com/earlephilhower/arduino-pico/releases/download/global/package_rp2040_index.json`

still the same error:
```/home/radzik/Arduino/E46-VA-Loc-Monitoring/E46-VA-Loc-Monitoring.ino:17:10: fatal error: hardware/powman.h: No such file or directory
   17 | #include "hardware/powman.h"
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
exit status 1

Compilation error: hardware/powman.h: No such file or directory
```

**Comment by maxgerhardt** [comment]
The PR is not even merged yet. The entire release cycle will take even longer, until a new stable version is released for easy download. The fastest way will be if you search for the `lib\core_inc.txt` file locally on your computer and make the needed modification to it locally.

**Comment by radek-turecki** [comment]
compiled successfully! 👍 !

Oooo it will be long night before me :-) Now I can continue ma project and try to decrease power consumptions...

Thanks a lot!

**Comment by maxgerhardt** [comment]
Just FYI, there was also a slightly more expanded version of a sketch / library in https://github.com/earlephilhower/arduino-pico/pull/2620 that turned off more stuff, and has some hints to VBUS/VSYS powering, that brought it down to 760µA at 3.3V.

**Comment by radek-turecki** [comment]
Hello,

It is very difficult to use powman, and I am not that much of an expert. Yesterday, I tried various combinations, but the results are very poor. I got it down to about 10mA, but that is still almost +50% of what the car consumes when idle :-( So the issue is still relevant and continues to bother me.

Bypassing the regulator – I left this for the next step, but first I would like to optimize it as much as possible at the software level.

In my humble opinion, ultimately it should be encapsulated in two simple functions, e.g. deepSleep4ms(ms) and deepSleep4Pin(GPIx), for sleeping at a specified time and on a specified pin, and that's it.

@maxgerhardt  - Could you help me and send me a small piece of code that uses the function you suggested:
`void sleep_goto_dormant_until(struct timespec *ts, aon_timer_alarm_handler_t callback);`

I would appreciate your help.

**Comment by seamusdemora** [comment]
Like the Phoenix, this project may be rising from the ashes?  But I think I'll wait a while to see how things pan out!

**Comment by radek-turecki** [comment]
@seamusdemora Does that mean the topic has been brought up many times but never actually implemented?

Darn, that’s too bad, because I have a Pico 2 board, but with that much power consumption, I’ll have to use a different one to save energy, and I’ll use the Pico 2 for another project.

Too bad :-(

**Comment by maxgerhardt** [comment]
My first tests on a Pimoroni Plasma 2350 board and PPK2 powering with 5V gives a regular idle current of about 18mA, which in DORMANT mode and my current GPIO configuration (all input then input gate disabled) drops to around 2.39mA. Still too much.

<img width="1877" height="833" alt="Image" src="https://github.com/user-attachments/assets/2a76f9fd-f202-4465-96a5-fd456a2dcde5" />

**Comment by maxgerhardt** [comment]
Mhm, setting the dormant clock source to the RP2350-only LPOSC improves things, bringing it to 934µA at 5V input. I still need to understand the clocks and power domains better.

<img width="1920" height="1040" alt="Image" src="https://github.com/user-attachments/assets/27026036-3f47-4efd-8a5c-7599a183d8bf" />

**Comment by maxgerhardt** [comment]
Using a different PowMan sleeping technique (https://github.com/mamba2410/rp2350-powman-sleep/) that actually powers off the entire CPU core but needs a more difficult register / state restore code, I get it down to 475µA at 5V, but still much higher than the claimed "<150µA". GPIO pin initialization has a major effect on current draw: Different floating input vs. driving high / low vs pullup / pull down settings are needed per-pin.

<img width="1920" height="1040" alt="Image" src="https://github.com/user-attachments/assets/ad3f5942-6206-4cec-9020-46a9ee706b71" />

**Comment by seamusdemora** [comment]
@radek-turecki 

> [@seamusdemora](https://github.com/seamusdemora) Does that mean the topic has been brought up many times but never actually implemented?
> 
> Darn, that’s too bad, because I have a Pico 2 board, but with that much power consumption, I’ll have to use a different one to save energy, and I’ll use the Pico 2 for another project.
> 
> Too bad :-(

You can see the October, 2024 dates at the top of this thread... there have been other (related) threads.  IIRC, there were similar threads for the RP2040.

**Comment by dquadros** [comment]
> [@seamusdemora](https://github.com/seamusdemora) Does that mean the topic has been brought up many times but never actually implemented?
> Darn, that’s too bad, because I have a Pico 2 board, but with that much power consumption, I’ll have to use a different one to save energy, and I’ll use the Pico 2 for another project.

Short answer: yes.

The documentation, libraries and examples fall short. 

Even if we get a good deep sleep implementation, the RP2350 and the Pico 2 are not aimed at low power projects. The numbers on the RP2350 datasheet suggest that you can lower power consume a lot, but it is not in the same league as microcontrollers designed for low power.

**Comment by earlephilhower** [comment]
I would suggest not letting "perfect" be the enemy of "good enough" in this case.  There's lots of "good enough" in this core, and we still get by somehow.

If someone here actually has a use for low power modes on the RP2350 (RP2040 low power is DOA) then I would love if they threw together a simple PR implementing "some kind" of API, whatever seems appropriate.  After there's something there, improvements and bug fixes can come from a larger group of folks (hopefully).

While I have built lots of other stuff, low power testing needs better equipment and setups than I've got available, so unfortunately I'm not the guy to propose something.  My 20 yr old Craftsman multimeter's ammeter isn't nearly accurate enough to look at the uA range without enormous error bars.

**Comment by maxgerhardt** [comment]
A month ago, RP2350 support was added to the Adafruit SleepyDog library (for watchdogs and sleeping):

https://github.com/adafruit/Adafruit_SleepyDog/tree/master/examples/SleepRP2350Timer

So I'm not sure if that's something that needs to be added to the Arduino core still.. RP2040 "support" is there too but only in the simplest form, a simple forward to Pico-SDK's `sleep_ms()` which might do a WFI.

**Comment by palmerr23** [comment]
I've tried the SleepyDog library and the saving for the basic Sleep example is only 2mA (19 ->17mA) on a Pico 2.

I can't get the more ambitious SleepRP2350Timer example to compile without an added:

#define ARDUINO_ARCH_RP2350

and the loader fails when it is included.

My guess is that it's still a (very welcome) work in progress.

**Comment by radek-turecki** [comment]
2mA - wow! Do you think that for the simple project it will be working and will be compiled?

**Comment by palmerr23** [comment]
There is a solution and PR in this thread [(https://forums.adafruit.com/viewtopic.php?t=223127)].

The SleepRP2350Timer example, after making the changes above, Sleep reduced the power from 19mA to just over 2mA, which is quite a worthwhile improvement! Dormant brings the current down further to a little lower than 2mA.

USB serial doesn't seem to re-establish properly after the sleep. While connection is still seen by Arduino's Serial Monitor, nothing comes through.

Unplug/replug after re-programming to correctly reset the system clock - otherwise the Pico 2 might not ever wake!

**Comment by radek-turecki** [comment]
Wow! Thank you so much! I'm going back to 2350 and test it. Thank you.

**Comment by radek-turecki** [comment]
I used example from:
https://github.com/adafruit/Adafruit_SleepyDog/blob/master/examples/SleepRP2350Timer/SleepRP2350Timer.ino

I added at the beginning of the code:
`#define ARDUINO_ARCH_RP2350`

and the result of compiler is :-( why?
```
/home/radzik/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/radzik/.cache/arduino/sketches/9071D2FE56B25DCBE9657D682305901C/sketch/Sleep_RP2350.ino.cpp.o: in function `loop':
/home/radzik/Arduino/Sleep_RP2350/Sleep_RP2350.ino:46:(.text.loop+0x28): undefined reference to `_ZN14WatchdogRP235014goToSleepUntilEib'
/home/radzik/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/radzik/Arduino/Sleep_RP2350/Sleep_RP2350.ino:58:(.text.loop+0x32): undefined reference to `_ZN14WatchdogRP235015resumeFromSleepEv'
/home/radzik/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/radzik/Arduino/Sleep_RP2350/Sleep_RP2350.ino:62:(.text.loop+0x48): undefined reference to `_ZN14WatchdogRP235016getSleepDurationEv'
/home/radzik/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/radzik/.cache/arduino/sketches/9071D2FE56B25DCBE9657D682305901C/sketch/Sleep_RP2350.ino.cpp.o: in function `setup':
/home/radzik/Arduino/Sleep_RP2350/Sleep_RP2350.ino:35:(.text.setup+0x32): undefined reference to `_ZN14WatchdogRP23509setWakeCbEPFvvE'
collect2: error: ld returned 1 exit status
exit status 1

Compilation error: exit status 1
```

**Comment by palmerr23** [comment]
You need to make all the changes in the PR `https://github.com/adafruit/Adafruit_SleepyDog/pull/64`

**Comment by radek-turecki** [comment]
I'm not a coding expert, unfortunately, but I understand that the issue has been resolved....

So should I wait until the new version for the Pico 2 to Mother Board is released? Or should I add all the #defines to the 'h' files?

**Comment by brentru** [comment]
We're testing this SleepyDog PR and will likely release it today.

> USB serial doesn't seem to re-establish properly after the sleep. While connection is still seen by Arduino's Serial Monitor, nothing comes through. Unplug/replug after re-programming to correctly reset the system clock - otherwise the Pico 2 might not ever wake!

(@earlephilhower  can correct me if I'm wrong), I believe that's due to how TinyUSB's CDC doesn't perform automatic re-enumeration after deep sleep.

**Comment by earlephilhower** [comment]
> ([@earlephilhower](https://github.com/earlephilhower) can correct me if I'm wrong), I believe that's due to how TinyUSB's CDC doesn't perform automatic re-enumeration after deep sleep.

The Pico doesn't enumerate, that's the hots's job.  If you're using the core USB stack (not AdaFruit TinyUSB) you can do a `USB.disconnect(); USB.connect();` after waking up to pretend to unplug/plug the device and signal the host to scan it.

But that is predicated on everything being restored in the USB block after wake up.  If the clocks aren't right or registers are reset, then all bets are off.

**Comment by palmerr23** [comment]
Thanks for the new RP2350 code Brent, it's scratched a long-felt itch!

A couple of improvements to the SleepRP2350Timer example:

After `Watchdog.resumeFromSleep();`
the following lines should be added to re-enable USB and Serial:

```
USB.disconnect(); 
USB.connect();
Serial.begin(115200);
    while(!Serial);
```

A note in the header that the device needs to be unplugged after uploading the code for the timer to be reset correctly.

The same is probably true for the other RP2350 example as well.

**Comment by palmerr23** [comment]
Adafruit is about to release v1.8.2 of their SleepyDog library with the revised RP2350 sleep code.

Expect around 2mA current draw while in sleep mode on an original Pi Pico 2  board, down from around 19mA while running.

Mischief (substantially) managed!

**Comment by earlephilhower** [comment]
@palmerr23 I suggest your put your fixes in a PR at the SleepyDog repo so will all get it.

**Comment by palmerr23** [comment]
Earle, 

USB and Serial restarts PR inclusion under discussion with Adafruit staff.  Apparently Mac and Windows behave differently.

**Comment by radek-turecki** [comment]
Wow! Wow! Wow!
1. Arduino IDE updated for SleepyDog 1.8.2
2. Compilation "from the finger" - without errors
3. After test on Waveshare RP2350B-Plus-W - in sleep mode ~2 mA

Brilliant!!!!! Good job!!! Thank you!!!

**Comment by brentru** [comment]
@radek-turecki  Thanks, glad it is working for you! I added this to SleepyDog a month+ ago for [Adafruit WipperSnapper firmware](https://github.com/adafruit/Adafruit_Wippersnapper_Arduino/). During dev., I performed ULP testing with a Nordic PPK II. I guess this suddenly has traction because Max mentioned it and I am happy it is working.

@palmerr23  Because SleepyDog is not part of the arduino-pico core, plz keep any issues related to SleepyDog to the [library's issues page](https://github.com/adafruit/Adafruit_SleepyDog/issues) or [pull request page](https://github.com/adafruit/Adafruit_SleepyDog/pulls), or the Adafruit Forums.

**Comment by radek-turecki** [comment]
Sure!
Once again BIG THANK YOU!

**Comment by tyeth** [comment]
Just to close this out by adding that the latest release (1.8.4) includes example fixes, and is tested with the various usb stack possibilities on RP2040+RP2350. We'll also now be auto-building for RP2 targets so future compilation errors can be avoided.

**Comment by radek-turecki** [comment]
Something wrong is again :-(

The code exactly like in the example:
https://github.com/adafruit/Adafruit_SleepyDog/blob/master/examples/SleepRP2350Timer/SleepRP2350Timer.ino

after command:
`Watchdog.goToSleepUntil(5000);`

after 5, 6, 7, ..., 1000sec the module is still sleeping :-(

**Comment by earlephilhower** [comment]
Did you completely unpower (unplug and no battery backup) the Pico as specified in the example?  I know I made that mistake and had the same issue, but unplugging and replugging made it work as expected.

**Comment by radek-turecki** [comment]
Hmmm OK! :-) It works! Perfect! Thank you so much @earlephilhower!

I didn't realize it could make such a big difference. I hope this restart is only required once after the upload, and that the module will ultimately be able to run for many days without restarting, with periodic sleep cycles.

Once again - thank you! Have a sunny day!

**Comment by radek-turecki** [comment]
I guess my excitement after the first attempt was a bit premature :-( Another round of programming -> upload -> USB reconnect, and... the module won't wake up :-(

The code is as simple as it gets:
```
awake = false;
  digitalWrite(LED_PIN, LOW);
  
  // Enter sleep state (6.5.2 in RP2350 Datasheet)
  Watchdog.goToSleepUntil(SLEEP_TIME_PERIOD); //30000ms
  // Uncomment the line below (and comment the line above) to enter Dormant
  // State (6.5.3 in RP2350 Datasheet) instead of the Sleep State above
  // Watchdog.goToSleepUntil(5000, true);

  // Make sure we don't wake
  /*
  while (!awake) {
    Serial.println("Should be sleeping here...");
  }
  */

  
  // Re-enable clocks, generators, USB and resume execution
  // NOTE: This MUST be called to properly resume from sleep!
  Watchdog.resumeFromSleep();
  Serial.end(); 
  delay(100);
  Serial.begin(115200);
  delay(100);
  //  while(!Serial);

  if (DEBUG) {
    PL("I'm awake now!");
    P("Slept for approximately ");
    long sleep_duration = Watchdog.getSleepDuration();
    Serial.print(sleep_duration);
    PL(" milliseconds.");
  }
  digitalWrite(LED_PIN, HIGH);
```
grrrr

**Comment by earlephilhower** [comment]
Your best bet is to post something in that library repo.  The authors there don't necessarily follow this repo or thread.

**Comment by palmerr23** [comment]
For others' reference, explicit reconnection to USB is required, then Serial() restarted.

This code works with SleepyDog > 1.8.4:
```
  Watchdog.goToSleepUntil(5000);
  // Uncomment the line below (and comment the line above) to enter Dormant
  // State (6.5.3 in RP2350 Datasheet) instead of the Sleep State above
  // Watchdog.goToSleepUntil(5000, true);

  // Make sure we don't wake
  while (!awake) {
    Serial.println("Should be sleeping here...");
  }  Watchdog.resumeFromSleep();
 USB.disconnect(); //delay(100); 
 USB.connect();
 Serial.begin(115200);
    while(!Serial);
```
See [https://forums.adafruit.com/viewtopic.php?p=1084737&hilit=sleepydog#p1084737](url)

**Comment by radek-turecki** [comment]
1. It seems void cbWake() need to be define i declare - it seems it must be obligatory. 
2. That is right - USB (output) it dosn't work after wake-up. USB.disconnect();/USB.connect(); - do you mean to use Adafruit_TinyUSB.h and USB stack Adafruit (insted of Pico SDK)?

**Comment by palmerr23** [comment]
Yes `cbWake( )` is required, it's the 'timer' callback after the delay finishes.

Mine just sets the awake flag so that `Watchdog.resumeFromSleep( );` executes.

```
void cbWake() {
  // Show we're awake again
  digitalWrite(LED_BUILTIN, HIGH);
  awake = true;
}
```
I use the USB stack that is packaged with Arduino-Pico as default.

The best place for this discussion is in the Arduino Forums thread for the SleepDog library. The engineers are quite responsive.

See [https://forums.adafruit.com/viewtopic.php?p=1084737&hilit=sleepydog#p1084737](https://github.com/earlephilhower/arduino-pico/issues/url)

**Comment by radek-turecki** [comment]
I got it after many tests and few hours later....

Based on my tests it seems that the problem with the resume is when the WiFi connection is established. So before the calling `Watchdog.goToSleepUntil(ms);`  there is necessary to put:
`WiFi.disconnect();`

and after resuming the module there is necessare to connect to WiFi again.
